lib/Runtime/Library/JavascriptBigInt.cpp

```diff
@@ -0,0 +1,215 @@
+    {
+        switch (JavascriptOperators::GetTypeId(right))
+        {
+        case TypeIds_BigInt:
+            *value = JavascriptBigInt::Equals(left, right);
+            break;
+        case TypeIds_Integer:
+        case TypeIds_Number:
+        case TypeIds_Int64Number:
+        case TypeIds_UInt64Number:
+        case TypeIds_Boolean:
+        case TypeIds_String:
+        case TypeIds_Symbol:
+        case TypeIds_VariantDate:
+        case TypeIds_Undefined:
+        case TypeIds_Null:
+        default:
+            AssertMsg(JavascriptBigInt::Is(right), "do not support comparison with types other than BigInt");
+            *value = FALSE;
+            break;
+        }
+        return true;
+    }
+
+    template <typename EncodedChar>
+    void JavascriptBigInt::InitFromCharDigits(const EncodedChar *pChar, uint32 charLength, bool isNegative)
+    {
+        Assert(charLength >= 0);
+        Assert(pChar != 0);
+        AssertMsg(charLength <= MaxStringLength, "do not support BigInt out of the range");
+
+        const EncodedChar *pCharLimit = pChar + charLength - 1;//'n' at the end
+
+        m_length = 0;
+        m_digits = m_reservedDigitsInit;
+        m_isNegative = isNegative;
+
+        uint32 digitMul = 1;
+        uint32 digitAdd = 0;
+        bool check = true;
+        for (; pChar < pCharLimit; pChar++)
+        {
+            Assert(NumberUtilities::IsDigit(*pChar));
+            if (digitMul == 1e9)
+            {
+                check = MulThenAdd(digitMul, digitAdd);
+                Assert(check);
+                digitMul = 1;
+                digitAdd = 0;
+            }
+            digitMul *= 10;
+            digitAdd = digitAdd * 10 + *pChar - '0';
+        }
+        Assert(1 < digitMul);
+        check = MulThenAdd(digitMul, digitAdd);
+        Assert(check);
+
+        // make sure this is no negative zero
+        if (m_length == 0)
+        {
+            m_isNegative = false;
+            m_length = 1;
+            m_digits[0] = 0;
+        }
+    }
+
+    bool JavascriptBigInt::MulThenAdd(uint32 digitMul, uint32 digitAdd)
+    {
+        Assert(digitMul != 0);
+
+        uint32 carryDigit = 0;
+        uint32 *pDigit = m_digits;
+        uint32 *pDigitLimit = pDigit + m_length;
+
+        for (; pDigit < pDigitLimit; pDigit++)
+        {
+            *pDigit = NumberUtilities::MulLu(*pDigit, digitMul, &carryDigit);// return low Digit to digit, hight Digit to carry
+            if (digitAdd > 0)
+            {
+                carryDigit += NumberUtilities::AddLu(pDigit, digitAdd);// add carry to result
+            }
+            digitAdd = carryDigit;
+        }
+        if (0 < digitAdd) // length increase by 1
+        {
+            AssertOrFailFast(m_length < MaxDigitLength);// check out of bound
+            m_digits[m_length++] = digitAdd;
+        }
+        return true;
+    }
+
+    int JavascriptBigInt::Compare(JavascriptBigInt *pbi)
```

Comment:
Some of the functions in this file are fairly intricate- if it's not too difficult, consider adding native unit tests to validate these individual functions

Reply:
thanks for your suggestion! I'm refactoring this part on the next PR. Will consider adding navtive tests in that PR.